samples/RoutingSample.Web/DomainPolicy/DomainMatcherPolicy.cs

```diff
@@ -0,0 +1,184 @@
+                var domains = endpoint.Metadata.GetMetadata<DomainMetadata>()?.Domains;
+                if (domains == null || domains.Count == 0)
+                {
+                    domains = new[] { new HostString() };
+                }
+
+                for (var j = 0; j < domains.Count; j++)
+                {
+                    var contentType = domains[j];
+
+                    if (!edges.ContainsKey(contentType))
+                    {
+                        edges.Add(contentType, new List<Endpoint>());
+                    }
+                }
+            }
+
+            // Now in a second loop, add endpoints to these lists. We've enumerated all of
+            // the states, so we want to see which states this endpoint matches.
+            for (var i = 0; i < endpoints.Count; i++)
+            {
+                var endpoint = endpoints[i];
+                var domains = endpoint.Metadata.GetMetadata<DomainMetadata>()?.Domains ?? Array.Empty<HostString>();
+                if (domains.Count == 0)
+                {
+                    // OK this means that this endpoint matches *all* domains.
+                    // So, loop and add it to all states.
+                    foreach (var kvp in edges)
+                    {
+                        kvp.Value.Add(endpoint);
+                    }
+                }
+                else
+                {
+                    // OK this endpoint matches specific domains
+                    foreach (var kvp in edges)
+                    {
+                        // The edgeKey maps to a possible request header value
+                        var edgeKey = kvp.Key;
+
+                        for (var j = 0; j < domains.Count; j++)
+                        {
+                            var domain = domains[j];
+
+                            if (edgeKey == domain)
+                            {
+                                kvp.Value.Add(endpoint);
+                                break;
+                            }
+                        }
+                    }
+                }
+            }
+
+            return edges
+                .Select(kvp => new PolicyNodeEdge(kvp.Key, kvp.Value))
+                .ToArray();
+        }
+
+        public PolicyJumpTable BuildJumpTable(int exitDestination, IReadOnlyList<PolicyJumpTableEdge> edges)
+        {
+            if (edges == null)
+            {
+                throw new ArgumentNullException(nameof(edges));
+            }
+
+            // Since our 'edges' can have wildcards, we do a sort based on how wildcard-ey they
+            // are then then execute them in linear order.
+            var ordered = edges
+                .Select(e => (domain: (HostString)e.State, destination: e.Destination))
+                .OrderBy(e => GetScore(e.domain))
+                .ToArray();
+
+            return new DomainPolicyJumpTable(exitDestination, ordered);
+        }
+
+        private int GetScore(in HostString domain)
```

Comment:
This probably deserves a comment about why it's not used by the comparer. It makes sense to me. But I might forget.

Reply:
Maybe this should be used by the comparer. See https://github.com/aspnet/Routing/pull/847/files#r224504940

Reply:
I copied this from the ConsumesMatcherPolicy. Should it be using the score in the comparer as well?